applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+    {
+        const std::string design_varible_name = this->GetValue( DESIGN_VARIABLE_NAME );
+
+        if (KratosComponents<Variable<double>>::Has(design_varible_name) == true)
+        {
+            const Variable<double>& r_variable =
+                KratosComponents<Variable<double>>::Get(design_varible_name);
+            this->CalculateStressDesignVariableDerivative(r_variable, STRESS_ON_GP, rOutput, rCurrentProcessInfo);
+        }
+        else if (KratosComponents<Variable<array_1d<double, 3>>>::Has(design_varible_name) == true)
+        {
+            const Variable<array_1d<double, 3>>& r_variable =
+                KratosComponents<Variable<array_1d<double, 3>>>::Get(design_varible_name);
+            this->CalculateStressDesignVariableDerivative(r_variable, STRESS_ON_GP, rOutput, rCurrentProcessInfo);
+        }
+    }
+    else if(rVariable == STRESS_DESIGN_DERIVATIVE_ON_NODE)
+    {
+        std::string design_varible_name = this->GetValue( DESIGN_VARIABLE_NAME );
+
+        if (KratosComponents<Variable<double>>::Has(design_varible_name) == true)
+        {
+            const Variable<double>& r_variable =
+                KratosComponents<Variable<double>>::Get(design_varible_name);
+            this->CalculateStressDesignVariableDerivative(r_variable, STRESS_ON_NODE, rOutput, rCurrentProcessInfo);
+        }
+        else if (KratosComponents<Variable<array_1d<double, 3>>>::Has(design_varible_name) == true)
+        {
+            const Variable<array_1d<double, 3>>& r_variable =
+                KratosComponents<Variable<array_1d<double, 3>>>::Get(design_varible_name);
+            this->CalculateStressDesignVariableDerivative(r_variable, STRESS_ON_NODE, rOutput, rCurrentProcessInfo);
+        }
+    }
+        else
+    {
+        rOutput.clear();
+    }
+
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferencingBaseElement::GetValueOnIntegrationPoints(const Variable<double>& rVariable,
+                    std::vector<double>& rValues,
+                    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(this->Has(rVariable))
+    {
+        // Get result value for output
+        double output_value = this->GetValue(rVariable);
+
+        // Resize Output
+        const unsigned int&  write_points_number = GetGeometry()
+            .IntegrationPointsNumber(this->GetIntegrationMethod());
+        if (rValues.size() != write_points_number)
+        {
+            rValues.resize(write_points_number);
+        }
+
+        // Write scalar result value on all Gauss-Points
+        for(unsigned int i = 0; i < write_points_number; ++i)
+        {
+            rValues[i] = output_value;
+        }
+    }
+    else
+        KRATOS_ERROR << "Unsupported output variable." << std::endl;
+
+    KRATOS_CATCH("")
+}
+
+int AdjointFiniteDifferencingBaseElement::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    GeometryType& r_geom = GetGeometry();
```

Comment:
I am not sure abt calling the check of the primals in `Check` of the Adjoints

If the primal analysis is also being executed then usually the Check will be called there

We should talk abt this

Reply:
mpPrimalElement->Check() does not work because the dofs are checked there.
For the adjoint problem, the dofs are not the same as the primal element expects, this is why this check would fail. Exceot from that i would also prefer to call the mpPrimalElement->Check(), because this is actually just a copy of the function there, with exchanged dofs.

Reply:
Ok, I will leave this. To explain my concern with copying the check instead of calling the primal's Check() is that such hard-coded synchronizations between classes break as soon as someone makes a local change to the primal element such as adding a variable. To avoid this one would need to refactor (also the primal elements).

Reply:
I would definitely also prefer to call the primal elements Check().
Unfortunately refactoring the primal elements Check() is also not enough, because I need to be able to call this via the public element interface...